README.md
# Mopidy-Spotify

[![Latest PyPI version](https://img.shields.io/pypi/v/Mopidy-Spotify)](https://pypi.org/project/Mopidy-Spotify/)
[![CI build status](https://img.shields.io/github/workflow/status/mopidy/mopidy-spotify/CI)](https://github.com/mopidy/mopidy-spotify/actions)
[![Test coverage](https://img.shields.io/codecov/c/gh/mopidy/mopidy-spotify)](https://codecov.io/gh/mopidy/mopidy-spotify)

[Mopidy](https://mopidy.com/) extension for playing music from [Spotify](https://www.spotify.com/).

## Status

Mopidy-Spotify is dependent on [`pyspotify`](https://pyspotify.readthedocs.io/en/latest/), a wrapper for Spotify's `libspotify` C library.
`libspotify` was [deprecated in 2015](https://github.com/mopidy/mopidy-spotify/issues/110) with no replacement.
It is unmaintained, functionally limited, and also now unavailable from the [Spotify developer site](https://developer.spotify.com/technologies/).
Where possible, we are [moving to use Spotify's Web API](https://github.com/mopidy/mopidy-spotify/issues/114) instead.
However, native playback is still only possible using `libspotify` and there is no official way for us to provide some Spotify features.

Limitations and/or bugs in `libspotify` currently result in missing/broken Mopidy-Spotify support for the following:

- Saving items to My Music ([#108](https://github.com/mopidy/mopidy-spotify/issues/108)) - possible via web API
- Podcasts ([#201](https://github.com/mopidy/mopidy-spotify/issues/201)) - unavailable
- Radio ([#9](https://github.com/mopidy/mopidy-spotify/issues/9)) - unavailable
- Spotify Connect ([#14](https://github.com/mopidy/mopidy-spotify/issues/14)) - unavailable

Working support for the following features is currently available:

- Playback
- Search
- Playlists (read-only)
- Top lists and Your Music (read-only)
- Lookup by URI

## Dependencies

- A Spotify Premium subscription. Mopidy-Spotify **will not** work with Spotify
  Free, just Spotify Premium.

- A non-Facebook Spotify username and password. If you created your account
  through Facebook you'll need to create a "device password" to be able to use
  Mopidy-Spotify. Go to http://www.spotify.com/account/set-device-password/,
  login with your Facebook account, and follow the instructions. However,
  sometimes that process can fail for users with Facebook logins, in which case
  you can create an app-specific password on Facebook by going to facebook.com >
  Settings > Security > App passwords > Generate app passwords, and generate one
  to use with Mopidy-Spotify.

- `libspotify` 12. The official C library from our `Unofficial
  libspotify archive <https://mopidy.github.io/libspotify-archive/>`_.
  The package is available as `libspotify12` from
  `apt.mopidy.com <http://apt.mopidy.com/>`__.

- `pyspotify` >= 2.0.5. The `libspotify` Python wrapper. The package is
  available as `python3-spotify` from apt.mopidy.com or `pyspotify` on PyPI.
  See https://pyspotify.readthedocs.io/en/latest/installation/ for how to install
  it and its dependencies on most platforms.

- `Mopidy` >= 3.0. The music server that Mopidy-Spotify extends.

If you install Mopidy-Spotify from apt.mopidy.com, AUR, or Homebrew, these
dependencies are installed automatically.

## Installation

### PyPI

~~~~~
sudo python3 -m pip install Mopidy-Spotify
~~~~~

### MacOS (Homebrew)

~~~~~
brew install mopidy/mopidy/mopidy-spotify
~~~~~

### Debian / Ubuntu

This extension is currently not packaged in Debian/Ubuntu, but you can install the `mopidy-spotify` Debian package from Mopidy's APT repository.

First, follow the instructions at [apt.mopidy.com](https://apt.mopidy.com) on how to setup Mopidy's APT repo on your system.

Then, install the `mopidy-spotify` package:

~~~~~
sudo apt install mopidy-spotify
~~~~~

### Arch (AUR)

~~~~~
yay -S mopidy-spotify
~~~~~

### Fedora

~~~~~
dnf install mopidy-spotify
~~~~~

See https://mopidy.com/ext/spotify/ for alternative installation methods.

## Configuration

Before starting Mopidy, you must add your Spotify Premium username and password
to your Mopidy configuration file and also visit
https://mopidy.com/ext/spotify/#authentication
to authorize this extension against your Spotify account:

~~~~~
  [spotify]
  username = alice
  password = secret
  client_id = ... client_id value you got from mopidy.com ...
  client_secret = ... client_secret value you got from mopidy.com ...
~~~~~

The following configuration values are available:

- `spotify/enabled`: If the Spotify extension should be enabled or not.
  Defaults to `true`.

- `spotify/username`: Your Spotify Premium username. You *must* provide this.

- `spotify/password`: Your Spotify Premium password. You *must* provide this.

- `spotify/client_id`: Your Spotify application client id. You *must* provide this.

- `spotify/client_secret`: Your Spotify application secret key. You *must* provide this.

- `spotify/bitrate`: Audio bitrate in kbps. `96`, `160`, or `320`.
  Defaults to `160`.

- `spotify/volume_normalization`: Whether volume normalization is active or
  not. Defaults to `true`.

- `spotify/timeout`: Seconds before giving up waiting for search results,
  etc. Defaults to `10`.

- `spotify/allow_cache`: Whether to allow caching. The cache is stored in a
  "spotify" directory within Mopidy's `core/cache_dir`. Defaults to `true`.

- `spotify/allow_network`: Whether to allow network access or not. Defaults
  to `true`.

- `spotify/allow_playlists`: Whether or not playlists should be exposed.
  Defaults to `true`.

- `spotify/search_album_count`: Maximum number of albums returned in search
  results. Number between 0 and 50. Defaults to 20.

- `spotify/search_artist_count`: Maximum number of artists returned in search
  results. Number between 0 and 50. Defaults to 10.

- `spotify/search_track_count`: Maximum number of tracks returned in search
  results. Number between 0 and 50. Defaults to 50.

- `spotify/toplist_countries`: Comma separated list of two letter ISO country
  codes to get toplists for. Defaults to blank, which is interpreted as all
  countries that Spotify is available in.

- `spotify/private_session`: Whether to use a private Spotify session. Turn
  on private session to disable sharing of played tracks with friends through
  the Spotify activity feed, Last.fm scrobbling, and Facebook. This only
  affects social sharing done by Spotify, not by other Mopidy extensions.
  Defaults to `false`.

## Resources

- [Source code](https://github.com/mopidy/mopidy-spotify)
- [Issue tracker](https://github.com/mopidy/mopidy-spotify/issues)
- [Changelog](https://github.com/mopidy/mopidy-spotify/releases)

## Credits

- Original author: [Stein Magnus Jodal](https://github.com/jodal)
- Current maintainer: [Stein Magnus Jodal](https://github.com/jodal)
- [Contributors](https://github.com/mopidy/mopidy-spotify/graphs/contributors)


CHANGES.rst
Changelog
=========

v3.1.0 (UNRELEASED)
-------------------

Feature release.

- Include the artists of each album in the search results. (PR: #118)

- Respect `spotify/timeout` setting when loading data from Spotify. (#129, PR:
  #139)

- Use OAuth to authenticate Spotify Web API requests, which has required
  authentication since 2017-05-29. The new config value `spotify/client_id` and
  `spotify/client_secret` must be set. Refer to the README for details. (#130,
  #142, PR: #143)

v3.0.0 (2016-02-15)
-------------------

Feature release.

- Require Mopidy 2.0.

- Minor adjustments to work with GStreamer 1.

v2.3.1 (2016-02-14)
-------------------

Bug fix release.

- Require Mopidy < 2 as Mopidy 2.0 breaks the audio API with the upgrade to
  GStreamer 1.

- Use the new Spotify Web API for search. Searching through libspotify has been
  discontinued and is not working anymore. (Fixes: #89)

- Note that search through the Spotify Web API doesn't return artists or date
  for albums. This also means that ``library.get_distinct()`` when given type
  ``albumartist`` or ``date`` and a query only returns an empty set.

- Emit a warning if config value ``spotify/search_album_count``,
  ``spotify/search_artist_count``, or ``spotify/search_track_count`` is greater
  than 50, and use 50 instead. 50 is the maximum value that the Spotify Web API
  allows. The maximum in the config schema is not changed to not break existing
  configs.

v2.3.0 (2016-02-06)
-------------------

Feature release.

- Ignore all audio data deliveries from libspotify when when a seek is in
  progress. This ensures that we don't deliver audio data from before the seek
  with timestamps from after the seek.

- Ignore duplicate end of track callbacks.

- Don't increase the audio buffer timestamp if the buffer is rejected by
  Mopidy. This caused audio buffers delivered after one or more rejected audio
  buffers to have too high timestamps.

- When changing tracks, block until Mopidy completes the appsrc URI change.
  Not blocking here might break gapless playback.

- Lookup of a playlist you're not subscribed to will now properly load all of
  the playlist's tracks. (Fixes: #81, PR: #82)

- Workaround teardown race outputing lots of short stack traces on Mopidy
  shutdown. (See #73 for details)

v2.2.0 (2015-11-15)
-------------------

Feature release.

- As Spotify now exposes your "Starred" playlist as a regular playlist, we no
  longer get your "Starred" playlist through the dedicated "Starred" API. This
  avoids your "Starred" playlist being returned twice. Lookup of
  ``spotify:user:<username>:starred`` URIs works as before. (Fixes: #71)

- Interpret album year "0" as unknown. (Fixes: #72)

v2.1.0 (2015-09-22)
-------------------

Feature release.

- Require Requests >= 2.0.

- Support using a proxy when connecting to Spotify. This was forgotten in the
  2.0 reimplementation. (Fixes: #68)

- Support using a proxy when accessing Spotify's web API to get cover and
  artist imagery.

v2.0.1 (2015-08-23)
-------------------

Bug fix release.

- Filter out ``None`` from ``library.get_distinct()`` return values. (Fixes:
  #63)

v2.0.0 (2015-08-11)
-------------------

Rewrite using pyspotify 2. Should have feature parity with Mopidy-Spotify 1.

**Config**

- Add ``spotify/volume_normalization`` config. (Fixes: #13)

- Add ``spotify/allow_network`` config which can be used to force
  Mopidy-Spotify to stay offline. This is mostly useful for testing during
  development.

- Add ``spotify/allow_playlists`` config which can be used to disable all
  access to playlists on the Spotify account. Useful where Mopidy is shared by
  multiple users. (Fixes: #25)

- Make maximum number of returned results configurable through
  ``spotify/search_album_count``, ``spotify/search_artist_count``, and
  ``spotify/search_track_count``.

- Add ``spotify/private_session`` config.

- Change ``spotify/toplist_countries`` default value to blank, which is now
  interpreted as all supported countries instead of no countries.

- Removed ``spotify/cache_dir`` and ``spotify/settings_dir`` config values. We
  now use a "spotify" directory in the ``core/cache_dir`` and
  ``core/data_dir`` directories defined in Mopidy's configuration.

- Add ``spotify/allow_cache`` config value to make it possible to disable
  caching.

**Browse**

- Add browsing of top albums and top artists, in additon to top tracks.

- Add browsing by current user's country, in addition to personal, global and
  per-country browsing.

- Add browsing of artists, which includes the artist's top tracks and albums.

- Update list of countries Spotify is available in and provides toplists for.

**Lookup**

- Adding an artist by URI will now first find all albums by the artist and
  then all tracks in the albums. This way, the returned tracks are grouped by
  album and they are sorted by track number. (Fixes: #7)

- When adding an artist by URI, all albums that are marked as "compilations"
  or where the album artist is "Various Artists" are now ignored. (Fixes: #5)

**Library**

- The library provider method ``get_distinct()`` is now supported. When called
  without a query, the tracks in the user's playlists is used as the data
  source. When called with a query, a Spotify search is used as the data
  source. This addition makes the library view in some notable MPD clients,
  like ncmpcpp, become quite fast and usable with Spotify. (Fixes: #50)

**Playback**

- If another Spotify client starts playback with the same account, we get a
  "play token lost" event. Previously, Mopidy-Spotify would unconditionally
  pause Mopidy playback if this happened. Now, we only pause playback if we're
  currently playing music from Spotify. (Fixes: #1)

v1.4.0 (2015-05-19)
-------------------

- Update to not use deprecated Mopidy audio APIs.

- Use strings and not ints for the model's date field. This is required for
  compatibility with the model validation added in Mopidy 1.1. (Fixes: #52)


mopidy_spotify/__init__.py
from __future__ import unicode_literals

import os

from mopidy import config, ext


__version__ = '3.1.0'


class Extension(ext.Extension):

    dist_name = 'Mopidy-Spotify'
    ext_name = 'spotify'
    version = __version__

    def get_default_config(self):
        conf_file = os.path.join(os.path.dirname(__file__), 'ext.conf')
        return config.read(conf_file)

    def get_config_schema(self):
        schema = super(Extension, self).get_config_schema()

        schema['username'] = config.String()
        schema['password'] = config.Secret()

        schema['client_id'] = config.String()
        schema['client_secret'] = config.Secret()

        schema['bitrate'] = config.Integer(choices=(96, 160, 320))
        schema['volume_normalization'] = config.Boolean()
        schema['private_session'] = config.Boolean()

        schema['timeout'] = config.Integer(minimum=0)

        schema['cache_dir'] = config.Deprecated()  # since 2.0
        schema['settings_dir'] = config.Deprecated()  # since 2.0

        schema['allow_cache'] = config.Boolean()
        schema['allow_network'] = config.Boolean()
        schema['allow_playlists'] = config.Boolean()

        schema['search_album_count'] = config.Integer(minimum=0, maximum=200)
        schema['search_artist_count'] = config.Integer(minimum=0, maximum=200)
        schema['search_track_count'] = config.Integer(minimum=0, maximum=200)

        schema['toplist_countries'] = config.List(optional=True)

        return schema

    def setup(self, registry):
        from mopidy_spotify.backend import SpotifyBackend

        registry.add('backend', SpotifyBackend)
